File: api/lua/msg.lua
---@meta _

---@alias CallbackId integer

---@alias Length
---| "Fill"
---| { FillPortion: integer }
---| "Shrink"
---| { Fixed: float }

---@alias Anchor
---| "Top"
---| "Bottom"
---| "Left"
---| "Right"
---| "TopRight"
---| "TopLeft"
---| "BottomRight"
---| "BottomLeft"

---@alias Alignment
---| "Start"
---| "Center"
---| "End"

---@class (exact) Padding
---@field top float
---@field bottom float
---@field right float
---@field left float

---@alias Msg _Msg

---@class (exact) _Msg
---@field NewWidget NewWidget

---@class (exact) NewWidget
---@field widget WidgetDefinition
---@field width integer
---@field height integer
---@field anchor Anchor

---@class (exact) WidgetDefinition
---@field Slider Slider?
---@field Column Column?
---@field Button Button?
---@field Text Text?

---@class (exact) Slider
---@field range_start float
---@field range_end float
---@field value CallbackId
---@field on_change CallbackId
---@field on_release CallbackId?
---@field width Length
---@field height integer
---@field step float

---@class (exact) Column
---@field spacing integer
---@field padding Padding
---@field width Length
---@field height Length
---@field max_width integer
---@field alignment Alignment
---@field children WidgetDefinition[]

---@class (exact) Button
---@field width Length
---@field height Length
---@field padding Padding
---@field child WidgetDefinition

---@class (exact) Text
---@field text string

----------------------------------------------------------------

---@class (exact) IncomingMsg
---@field CallCallback CallCallback?

---@class (exact) CallCallback
---@field callback_id CallbackId
---@field args Args?

---@class Args

File: src/api/msg.rs
#[derive(Debug, serde::Serialize, serde::Deserialize, Clone, Copy)]
pub struct CallbackId(pub u32);

#[derive(Debug, serde::Serialize, serde::Deserialize, Clone, Copy)]
pub enum Anchor {
    Top,
    Bottom,
    Left,
    Right,
    TopRight,
    TopLeft,
    BottomRight,
    BottomLeft,
}

#[derive(Debug, serde::Deserialize)]
pub enum Msg {
    NewWidget {
        widget: WidgetDefinition,
        width: u32,
        height: u32,
        anchor: Anchor,
    },
}

#[derive(Debug, serde::Serialize)]
pub enum OutgoingMsg {
    CallCallback {
        callback_id: CallbackId,
        #[serde(default)]
        args: Option<Args>,
    },
}

#[derive(Debug, serde::Deserialize)]
pub enum WidgetDefinition {
    Slider {
        range_start: f32,
        range_end: f32,
        // value: ,
        on_change: CallbackId,
        #[serde(default)]
        on_release: Option<CallbackId>,
        #[serde(with = "LengthDef")]
        width: iced::Length,
        height: u16,
        step: f32,
    },
    Column {
        spacing: u16,
        #[serde(with = "PaddingDef")]
        padding: iced::Padding,
        #[serde(with = "LengthDef")]
        width: iced::Length,
        #[serde(with = "LengthDef")]
        height: iced::Length,
        max_width: u16,
        #[serde(with = "AlignmentDef")]
        alignment: iced::Alignment,
        children: Vec<WidgetDefinition>,
    },
    Button {
        #[serde(with = "LengthDef")]
        width: iced::Length,
        #[serde(with = "LengthDef")]
        height: iced::Length,
        #[serde(with = "PaddingDef")]
        padding: iced::Padding,
        child: Box<WidgetDefinition>,
    },
    Text {
        text: String,
        // size: u16,
        // line_height: TODO:
    },
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
#[serde(remote = "iced::Alignment")]
enum AlignmentDef {
    Start,
    Center,
    End,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
#[serde(remote = "iced::Padding")]
struct PaddingDef {
    top: f32,
    right: f32,
    bottom: f32,
    left: f32,
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
#[serde(remote = "iced::Length")]
enum LengthDef {
    Fill,
    FillPortion(u16),
    Shrink,
    Fixed(f32),
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub enum Args {
    SliderValue(f32),
}
